Replace hide_cols_by_index with hide_cols_by_label

fabiopipitone opened the pull request:
I introduced the possibility to hide a column by its label (simply specifying the labels separated by commas and without any superscript) and removed the possibility to hide it by its index since IMO it might be incredibly dangerous (if you use the split_by_cols feature, most likely you will have different number and types of columns depending on the global timerange filter). 

For the same reason, I added a try catch since if someone tries to hide a column that doesn't exist, I simply want the table to do nothing, not to throw an error and not showing anything (if I want to hide the column 'fabio' and I set a timerange where no event with field 'firstname'=='fabio' is present, I do not want the visualization to explode, but simply to do nothing about it). 

I obviously replaced the HiddenColumns placeholder, too.  

Personally, I suggest you should remove the possibility to directly reference a column by its index wholly, since it might be dangerous to do it even in the "Add computed columns" section.

fbaligand commented:
Hi @fabiopipitone,

First, thanks to contribute to enhanced-table plugin!

Then, concerning the PR:
- ability to hide a column by its label is a really nice feature!
- I planned to add it, so I'm happy that you did it!
- however, "hide a column by its label" should be a added feature, not a replacement.
- I mean: hidden columns should be hide-able by index or label
- It is important so that to not have a breaking change, and so, keep all existing visualizations still working. Thus, some people can prefer to hide a column by its id. So user should have choice.
- So the sample syntax could be: 0,1,'My Label',3

=> Could you review your code to support both column references by label and by index?

fabiopipitone commented:
Hi there!

Enhanced table is a beautiful plugin and I'm more than happy to contribute.

I guess I can (and will) adapt the code to keep both the features even
though I still cannot see the utility of hiding a column by its index
rather than its label.

I understand your compatibility claim but I think it'd be of more help to
enlighten the possibly unaware users of this HUGE risk and prevent them
from taking it (for no actual reason).

If I want to hide a column I want to hide a SPECIFIC column and not a
random one based on a random id.

If I were you, if not introducing the breaking change point blank, I would
at least warn the users in the comment of the release, strongly suggesting
them to use the hide by label feature because of the random behavior of
using the id. I may even start telling them the hide_col_by_id feature
might be removed in a future release (so they can start replacing it in
current visualizations).

Anyway, since as you said it might be annoying to change dozens of
visualizations for a user, I'll check it out tomorrow if I have enough time
and let you know.

Thank you for the feedback!

BR
Fabio

Il dom 8 set 2019, 23:26 Fabien Baligand <notifications@github.com> ha
scritto:

> Hi @fabiopipitone <https://github.com/fabiopipitone>,
>
> First, thanks to contribute to enhanced-table plugin!
>
> Then, concerning the PR:
>
>    - ability to hide a column by its label is a really nice feature!
>    - I planned to add it, so I'm happy that you did it!
>    - however, "hide a column by its label" should be a added feature, not
>    a replacement.
>    - I mean: hidden columns should be hide-able by index or label
>    - It is important so that to not have a breaking change, and so, keep
>    all existing visualizations still working. Thus, some people can prefer to
>    hide a column by its id. So user should have choice.
>    - So the sample syntax could be: 0,1,'My Label',3
>
> => Could you review your code to support both column references by label
> and by index?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/fbaligand/kibana-enhanced-table/pull/65?email_source=notifications&email_token=AK3SNTEJ3ICZNEGSQ7UZLU3QIVUXFA5CNFSM4IUKDQCKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD6FZYTA#issuecomment-529243212>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AK3SNTCONXHZKIP6U5MH6HLQIVUXFANCNFSM4IUKDQCA>
> .
>

fbaligand commented:
Hi,

Thanks for your comprehension and the change you plan.

Concerning "hidden col by index", I understand it can be random if you use "Split Cols" feature.
But if you don't use "Split Cols" feature, is there some risk?

fabiopipitone commented:
Hi,

it doesn't come to mind a possible dangerous situation in using hide cols
by id without using split cols (but I'll test it tomorrow and let you know).

Though, IMO it is better to give users a feature which work independently
from another previous setting (if the two aren't supposed to be linked like
here obviously).

Otherwise a beginner or simply a distracted user might not use it properly
and get back (heavily) wrong results.

If I was (I am in fact) a enhanced-table user, I'd rather spend some time
rapidly changing even 50 visualizations than risk buggy and unexpected
behaviors (maybe in production).

For the same reason I'd change the Computed Cols behavior, only allowing
the reference by label. I personally do not like potential random behaviors
and always try to picture the worst scenario.

Also, the throwing of the exception bypass is another important part.
I wouldn't risk not to visualize a table only because of a different
timerange setting.

Let me know what you think about it, too.

Bye,
Fabio

Il lun 9 set 2019, 00:02 Fabien Baligand <notifications@github.com> ha
scritto:

> Hi,
>
> Thanks for your comprehension and the change you plan.
>
> Concerning "hidden col by index", I understand it can be random if you use
> "Split Cols" feature.
> But if you don't use "Split Cols" feature, is there some risk?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/fbaligand/kibana-enhanced-table/pull/65?email_source=notifications&email_token=AK3SNTHLDGHI5IHUGSIQNCDQIVY6TA5CNFSM4IUKDQCKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD6F2MTQ#issuecomment-529245774>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AK3SNTAFWMYXPEOY4KKWGF3QIVY6TANCNFSM4IUKDQCA>
> .
>

fabiopipitone commented:
Hi there!

Ok, I've put the hide_column_by_id feature back alongside the new
hide_column_by_label.

Nevertheless, because of the riskiness of such a feature, I strongly advise
you to put a Deprecation Warning both in the release comment and below the
"Hidden Columns" input in the form, to suggest the users should use the
safer hide_col_by_label feature and to warn them about a possible future
removal (in a major release of the plugin) of such feature.

Also, I tested the plugin even inserting stuff in random order (e.g.
1,2,name1,3,name2) and it seems to work. Though, I'd advise the users not
to mix the two syntaxes (either you use the indeces - strongly unadvised -
or you use the names).

Let me know what you think about it and, obviously, if you find any
possible bug or problem.

BR,
Fabio

Il giorno lun 9 set 2019 alle ore 00:14 F. P. <fabio.pipitone93@gmail.com>
ha scritto:

> Hi,
>
> it doesn't come to mind a possible dangerous situation in using hide cols
> by id without using split cols (but I'll test it tomorrow and let you know).
>
> Though, IMO it is better to give users a feature which work independently
> from another previous setting (if the two aren't supposed to be linked like
> here obviously).
>
> Otherwise a beginner or simply a distracted user might not use it properly
> and get back (heavily) wrong results.
>
> If I was (I am in fact) a enhanced-table user, I'd rather spend some time
> rapidly changing even 50 visualizations than risk buggy and unexpected
> behaviors (maybe in production).
>
> For the same reason I'd change the Computed Cols behavior, only allowing
> the reference by label. I personally do not like potential random behaviors
> and always try to picture the worst scenario.
>
> Also, the throwing of the exception bypass is another important part.
> I wouldn't risk not to visualize a table only because of a different
> timerange setting.
>
> Let me know what you think about it, too.
>
> Bye,
> Fabio
>
> Il lun 9 set 2019, 00:02 Fabien Baligand <notifications@github.com> ha
> scritto:
>
>> Hi,
>>
>> Thanks for your comprehension and the change you plan.
>>
>> Concerning "hidden col by index", I understand it can be random if you
>> use "Split Cols" feature.
>> But if you don't use "Split Cols" feature, is there some risk?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/fbaligand/kibana-enhanced-table/pull/65?email_source=notifications&email_token=AK3SNTHLDGHI5IHUGSIQNCDQIVY6TA5CNFSM4IUKDQCKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD6F2MTQ#issuecomment-529245774>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AK3SNTAFWMYXPEOY4KKWGF3QIVY6TANCNFSM4IUKDQCA>
>> .
>>
>

fbaligand commented:
OK.

Concerning the fact to not raise an error if label does not exist, I'm OK with that on hidden columns. 
It is coherent with reference by index, where a too high index (higher than columns count) has no impact, and so, is ignored.

Concerning warning about reference by index, you can add a warning in README.md, in "Features" section. Something like:
- Hide some table columns (ex: 'Status',0 hides columns labeled 'Status' and at index 0)
-- It is recommended to use reference by label to avoid inconsistent results

fbaligand commented on public/enhanced-table-vis-params.html:
```diff
@@ -191,7 +191,7 @@
     <div class="form-group">
       <label>Hidden columns</label>
-      <input type="text" ng-model="editorState.params.hiddenColumns" class="form-control" placeholder="0,1,..." />
+      <input type="text" ng-model="editorState.params.hiddenColumns" class="form-control" placeholder="1,2,column_name_X..." />
```
Here, can you set `placeholder="0,1,column_label_X,..."`

fabiopipitone replied:
Sure. Will do asap

fbaligand commented on public/enhanced-table-vis-controller.js:
```diff
@@ -308,12 +308,30 @@ module.controller('EnhancedTableVisController', function ($scope, Private, confi
+          if (parseInt(item).toString() == item) {
+            let index = getRealColIndex(parseInt(item), splitColIndex);
+            let colToRemove = index - removedCounter;
+            console.log(colToRemove)
```
I understand that you use console.log during debug, but please remove them before publishing code.

fabiopipitone replied:
So sorry man. I pushed in a rush at work today. That's why code revision is fundamental :).
I'll remove all the console.logs in the next commit

fbaligand commented on public/enhanced-table-vis-controller.js:
```diff
@@ -308,12 +308,30 @@ module.controller('EnhancedTableVisController', function ($scope, Private, confi
+            let itemIndex = findColIndexByTitle(table.columns, item, item, splitColIndex);
+            let index = getRealColIndex(itemIndex, splitColIndex);
+            let colToRemove = index;
+            console.log(colToRemove)
```
Same remark for console.log

fabiopipitone replied:
Same as above. Gonna be removed

fbaligand commented on public/enhanced-table-vis-controller.js:
```diff
@@ -308,12 +308,30 @@ module.controller('EnhancedTableVisController', function ($scope, Private, confi
-        });
-        removedCounter++;
+        try {
+          if (parseInt(item).toString() == item) {
```
Here, could you use : `isNaN(parseInt(item))`

fabiopipitone replied:
I actually can't because if someone has a column named "1-whatever column", because of how JS works, it's not gonna return a NaN if you convert into an integer. It's gonna return 1.
That's why I had to do that horrible stuff. Should you have another way to accomplish such a task let me know

fbaligand replied:
OK.
So another way:
`if (item.matches(/ *[0-9]+ */) {`

fabiopipitone replied:
Regex looks good. We're talking of elasticsearch after all ;)
Though, I think I'll leave it in a dedicate function `isInt()` to make the code more readable.

fbaligand replied:
Nice.

fbaligand commented on public/enhanced-table-vis-controller.js:
```diff
@@ -308,12 +308,30 @@ module.controller('EnhancedTableVisController', function ($scope, Private, confi
+            table.columns.splice(colToRemove, 1);
+            _.forEach(table.rows, function (row) {
+              row.splice(colToRemove, 1);
+            });
```
These 4 lines are identical in the `if` and `else` sections.
Can you put this code after if/else statements to avoid duplicate code?

fabiopipitone replied:
I can but then I'll have to add a duplicate of `(parseInt(item).toString() == item)` to decide whether to increase `removedCounter` or not. Ideas?

fabiopipitone replied:
Actually I just found a little bug in using a mix of column indeces and column labels. 
In short, removedCounter++ should be kept in both cases but it can create some problems if you remove a column "test_column" (which has index X, say 2) and then you remove column with index Y<X, say 1. In that case, removedCounter would be increased to 1 at the time of removing "test_column". Thus, when it comes to remove column at index 1, it'll remove column at index colToRemove = 1 - 1 = 0, which is wrong. 
Though, it is a potential bug already present in the original release and it's due to the fact of not putting the column in order (if you write "4,5,2,3" in Hidden Columns, you'll see the behavior won't be the expected one). 
I'll see what I can do about it tomorrow. 
In case we do not find an elegant solution we can always put a disclaimer which specifies to use EITHER the column indeces OR the column_labels and, in case someone wants to use the indeces, to write them in ascending order.

fabiopipitone replied:
Ok this should be fixed.

fbaligand commented on README.md:
```diff
@@ -19,7 +19,10 @@ This Kibana visualization plugin is like a Data Table, but with enhanced feature
+- Hide some table columns (ex: `0,1,My_Column` hides columns 0, 1 and the column labeled 'My_Column')
+  - Note that the column label must be written as is (including whitespaces), without any superscript/apostrophe.
+  - Column labels containing commas are not supported since the comma is used to separate the columns
+  - It is strongly recommended to use column labels to hide columns rather than their indeces. Using the Split By Cols feature, in fact, indeces might   depend on the global timerange that has been set. 
```
replace 'indeces' by 'index'
replace *Split By Cols feature* by *'Split cols' feature*

fabiopipitone replied:
Actually, I put the plural (with a typo since it should be `indices`) since it sounds better (`labels` is plural too). I can change it if you want but it would sound strange (like saying "it's recommended to use cars rather than bicycles when it rains", you wouldn't say "bicycle").

I'll definitely change `Split By cols` into `'Split cols'`

fbaligand replied:
OK for indices.

fabiopipitone commented on public/enhanced-table-vis-params.html:
```diff
@@ -191,7 +191,7 @@
     <div class="form-group">
       <label>Hidden columns</label>
-      <input type="text" ng-model="editorState.params.hiddenColumns" class="form-control" placeholder="1,2,column_name_X..." />
+      <input type="text" ng-model="editorState.params.hiddenColumns" class="form-control" placeholder="0,1,2,column_label_X..." />
```
Good idea. I'll add change it right now.

fabiopipitone commented:
I added regex inside the `isInt(item)' function. 
I added a brief description near the form input about how to use the `Hidden columns` feature so the users who haven't read the README can understand it better. 
Finally, I noticed there was a buggy behavior if you tried to hide the same column twice (e.g. 1,1 or 1,col_1_label) and I fixed it.

fbaligand commented on public/enhanced-table-vis-controller.js:
```diff
@@ -306,7 +306,7 @@ module.controller('EnhancedTableVisController', function ($scope, Private, confi
 
   const isInt = (item) => {
-    return parseInt(item).toString() == item;
+    return /^[0-9]+$/.test(item);
```
It would be nice to support leading and trailing spaces.
So can you put `/^ *[0-9] *$/` ?
IMHO, in previous versions, this was supported.
I don't want that `' 2 '` is interpreted as a label.

fabiopipitone replied:
Absolutely. I didn't mean to remove the trailing spaces. I'll fix it right now.

fbaligand commented on public/enhanced-table-vis-params.html:
```diff
@@ -190,8 +190,14 @@
+      <label>
+        Hidden columns
+        <icon-tip
+          content="'A column can be referenced by its index (e.g. 1,2,3..) or by its label (e.g. example_column,Example Column). It is important not to put any superscript or inverted comma if they are not part of the column label and to separate the columns to hide simply using a comma (labels containing commas cannot be hidden). References by id and by label can be mixed (e.g. 1,2,ExampleColumn ...
```
Nice to add a icon tip.
Documentation is always nice, but can't this description be a little bit shorter? :)

fabiopipitone replied:
No problem ;)

fbaligand commented on public/enhanced-table-vis-params.html:
```diff
@@ -190,8 +190,14 @@
+          position="'right'"
+        ></icon-tip>
+      </label>
+      <input type="text" ng-model="editorState.params.hiddenColumns" class="form-control" placeholder="0,1,2,Col2 Label,..." />
```
As you set `Col2 Label`, can you remove `2` index?

fabiopipitone replied:
Sure. I missed that one. Thank you

fabiopipitone commented:
Name of the commit should have been "Remove `STRONGLY` from the 'Hide columns' tool tip" but it didn't get `STRONGLY` since I used backticks rather than superscripts. 

Anyway, it should be fine now ;)

fabiopipitone commented:
Done. And I added the suggested fixes about stuff I made. 
Though, it still returns some warnings about things I didn't change, so I didn't touch them. 

Basically, it complains about a missing license header in various files:

```
yarn run v1.17.3
$ ../../kibana/node_modules/.bin/eslint .
                                                                                                                    
/home/fabio/Documenti/KP_Work/ELK/enhanced_table_test_2/kibana/kibana-enhanced-table/index.js                       
  1:1  error  File must start with a license header  @kbn/eslint/require-license-header                             
                                                                                                                    
/home/fabio/Documenti/KP_Work/ELK/enhanced_table_test_2/kibana/kibana-enhanced-table/public/agg_table/agg_table.js  
  148:15  error  Move function declaration to function body root  no-inner-declarations                             
                                                                                                                    
/home/fabio/Documenti/KP_Work/ELK/enhanced_table_test_2/kibana/kibana-enhanced-table/public/enhanced-table-request-handler.js
  1:1  error  File must start with a license header  @kbn/eslint/require-license-header

/home/fabio/Documenti/KP_Work/ELK/enhanced_table_test_2/kibana/kibana-enhanced-table/public/enhanced-table-vis-controller.js
  1:1  error  File must start with a license header  @kbn/eslint/require-license-header

/home/fabio/Documenti/KP_Work/ELK/enhanced_table_test_2/kibana/kibana-enhanced-table/public/enhanced-table-vis-hack.js
  1:1  error  File must start with a license header  @kbn/eslint/require-license-header

/home/fabio/Documenti/KP_Work/ELK/enhanced_table_test_2/kibana/kibana-enhanced-table/public/enhanced-table-vis-params.js
  1:1  error  File must start with a license header  @kbn/eslint/require-license-header

/home/fabio/Documenti/KP_Work/ELK/enhanced_table_test_2/kibana/kibana-enhanced-table/public/enhanced-table-vis.js
  1:1  error  File must start with a license header  @kbn/eslint/require-license-header

✖ 7 problems (7 errors, 0 warnings)
  6 errors and 0 warnings potentially fixable with the `--fix` option.

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

fbaligand commented:
OK.
I'm surprised about the license warning, I don't reproduce it.
Anyway, this is OK to me, there is no lint issue introduced by your code.

fbaligand commented:
So I can officially announce that your code is ready to merge !
**LGTM**
